src/daemon/rpmostreed-transaction-livefs.c

```diff
@@ -0,0 +1,650 @@
+   * our medium term goal here is to obviate most of the unlock usage.
+   */
+  OstreeDeploymentUnlockedState unlockstate = ostree_deployment_get_unlocked (booted_deployment);
+  if (unlockstate != OSTREE_DEPLOYMENT_UNLOCKED_NONE)
+    return glnx_throw (error, "livefs is incompatible with unlocked state");
+
+  /* Find the source for /etc - either booted or pending, but down below we
+     require pending */
+  OstreeDeployment *origin_merge_deployment =
+    rpmostree_syscore_get_origin_merge_deployment (sysroot,
+                                                   ostree_deployment_get_osname (booted_deployment));
+  g_assert (origin_merge_deployment);
+  const char *booted_csum = ostree_deployment_get_csum (booted_deployment);
+  const char *target_csum = ostree_deployment_get_csum (origin_merge_deployment);
+
+  /* Require a pending deployment to use as a source - perhaps in the future we
+   * handle direct live overlays.
+   */
+  if (origin_merge_deployment == booted_deployment)
+    return glnx_throw (error, "No pending deployment");
+
+  /* Open a fd for the booted deployment */
+  g_autofree char *deployment_path = ostree_sysroot_get_deployment_dirpath (sysroot, booted_deployment);
+  glnx_fd_close int deployment_dfd = -1;
+  if (!glnx_opendirat (ostree_sysroot_get_fd (sysroot), deployment_path, TRUE,
+                       &deployment_dfd, error))
+    return FALSE;
+
+  /* Find out whether we already have a live overlay */
+  g_autofree char *live_inprogress = NULL;
+  g_autofree char *live_replaced = NULL;
+  if (!rpmostree_syscore_deployment_get_live (sysroot, booted_deployment, -1,
+                                              &live_inprogress, &live_replaced,
+                                              error))
+    return FALSE;
+  if (live_inprogress != NULL)
+    {
+      if (strcmp (live_inprogress, target_csum) == 0)
+        g_print ("Note: Resuming interrupted overlay of %s\n", target_csum);
+    }
+  if (live_replaced != NULL)
+    {
+      if (strcmp (live_replaced, target_csum) == 0)
+        return glnx_throw (error, "Current overlay is already %s", target_csum);
+      g_print ("Note: Adding to previous overlay: %s\n", live_replaced);
+    }
+
+  /* Look at the difference between the two commits - we could also walk the
+   * filesystem, but doing it at the ostree level is potentially faster, since
+   * we know when two directories are the same.
+   */
+  g_autoptr(CommitDiff) diff = NULL;
+  if (!analyze_commit_diff (repo, booted_csum, target_csum,
+                            &diff, cancellable, error))
+    return FALSE;
+
+  print_commit_diff (diff);
+
+  const gboolean replacing = (self->flags & RPMOSTREE_TRANSACTION_LIVEFS_FLAG_REPLACE) > 0;
+  const gboolean requires_etc_merge = (diff->flags & COMMIT_DIFF_FLAGS_ETC) > 0;
+  const gboolean adds_packages = diff->added_pkgs->len > 0;
+  const gboolean modifies_packages = diff->removed_pkgs->len > 0 || diff->modified_pkgs_new->len > 0;
+  if (!adds_packages)
```

Comment:
Would it be more flexible if we weren't package-based at all for the new tree case? I.e. just check if it's a net file addition or if there are mutations? E.g. it would allow us to also handle postprocess scripts better.